docs/proposals/event-notification.md

```diff
@@ -0,0 +1,141 @@
+- Since these application servers are built to serve multiple concurrently-deployed applications, it is more difficult to determine the status of any specific application running on them. The lifecycle of the application server differs from the lifecycle of the application running inside the application server. 
+- Fortunately, in these cases we can use the console logs (from `odo log`) from the runtime container to determine a more detailed application status.
+- For example, OpenLiberty (as an example of an application-server-style container) prints a specific code when an application is starting  `CWWKZ0018I: Starting application {0}.`, and another when it has started. `CWWKZ0001I: Application {0} started in {1} seconds.`
+
+**Devfile writer**: A devfile writer may be a runtime developer (for example, a Red-Hatter working on WildFly or an IBMer working on OpenLibery) creating a devfile for their organization's runtime (for example 'OpenLiberty w/ Maven' dev file), or an application developer creating/customizing a dev file for use with their own application. In either case, the devfile writer must be familiar with the semantics of both odo and the devfile.
+
+## JSON-based odo command behaviours to detect app and build status
+
+New odo commands and flags:
+- `odo push -o json`
+- `odo component status -o json`
+
+With these two additions, an IDE or similar external tool can detect build running/succeeded/failed, application starting/started/stopping/stopped, and (in many cases) get a 'detailed app status' and/or 'detailed build status'.
+
+### Build status notification via `odo push -o json`
+
+`odo push -o json` is like standard `odo push`, but instead it outputs JSON events (including action console output) instead of text. This allows the internal state of the odo push process to be more easily consumed by external tools.
+
+Several different types of JSON-formatted events would be output, which correspond to odo container command/action executions:
+- Dev file command execution begun *(command name, start timestamp)*
+- Dev file command execution completed *(error code, end timestamp)*
+- Dev file action execution begun *(action name, parent command name, start timestamp)*
+- Dev file action execution completed *(action name, error code, end timestamp)*
+- Log/console stdout from the actions, one line at a time *(for example, `mvn build` output).* (timestamp)
+  
+(Exact details for which fields are included with events are TBD)
+
+In addition, `odo push -o json` should return a non-zero error code if one of the actions returned a non-zero error code, otherwise zero is returned.
+
+### `odo push -o json` example output
+
+This is what an `odo push -o json` command invocation would look like:
+```
+odo push -o json
+{ "devFileCommandExecutionBegun": { "commandName" : "build", "timestamp" : "(UTC unix epoch seconds.microseconds)" } }
```

Comment:
Each line has a different JSON structure (different keys), this might make parsing in some languages (like Go) a little bit trickier. Not a huge problem, but just something to be aware of.

Reply:
This is true; here's my draft PR of part of this feature (the PR has gone through zero review, so is subject to change) that contains the go structs used to marshal/unmarshall these entries: https://github.com/openshift/odo/blob/e8040649268c0e7624c5662e364fdb8607837b41/pkg/machineoutput/types_event_logging.go#L24